File: LeetCode/java/139_Word_Break.java
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    private boolean check(String s, Set<String> set, Boolean[] memo, int start) {
        if (start == s.length()) {
            return true;
        }

        if (memo[start] != null) {
            return memo[start];
        }

        for (int end = start + 1; end <= s.length(); end++) {
            String str = s.substring(start, end);

            if (set.contains(str)) {
                memo[end] = check(s, set, memo, end);

                if (memo[end] == true) {
                    return memo[end];
                }
            }
        }

        return false;
    }

    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> set = new HashSet<>(wordDict);
        Boolean[] memo = new Boolean[s.length() + 1];

        return check(s, set, memo, 0);
    }
}

File: LeetCode/cpp/139_Word_Break.cpp
class Solution {
public:
	bool check(string s, set<string>& st, vector<int>& memo, int start) {
		if (start == s.length()) {
			return true;
		}

		if (memo[start] != -1) {
			return memo[start];
		}

		for (int end = start + 1; end <= s.length(); end++) {
			string str = s.substr(start, end - start);

			if (st.find(str) != st.end()) {
				memo[end] = check(s, st, memo, end);

				if (memo[end] == 1) {
					return memo[end];
				}
			}
		}

		return false;
	}

	bool wordBreak(string s, vector<string>& wordDict) {
		set<string> st(wordDict.begin(), wordDict.end());
		vector<int> memo(s.length() + 1, -1);

		return check(s, st, memo, 0);
	}
};